jobrunner/run.py

```diff
@@ -195,28 +164,160 @@ def handle_running_job(job):
+            set_message(
+                job, "Waiting on dependencies", code=StatusCode.WAITING_ON_DEPENDENCIES
+            )
+            return
+
+        expected_state = ExecutorState.PREPARING
+        new_status = api.prepare(definition)
+
+    elif initial_status.state == ExecutorState.PREPARED:
+        expected_state = ExecutorState.EXECUTING
+        new_status = api.execute(definition)
+
+    elif initial_status.state == ExecutorState.EXECUTED:
+        expected_state = ExecutorState.FINALIZING
+        new_status = api.finalize(definition)
+
+    elif initial_status.state == ExecutorState.FINALIZED:
+        # final state - we have finished!
+        results = api.get_results(definition)
+        # TODO: implement workspace API
+        # delete_obsolete_files(job, results)
+        save_results(job, results)
+        mark_job_as_completed(job)
+        api.cleanup(definition)
+        # we are done here
+        return
+
+    # following logic is common to all non-final transitions
+
+    if new_status.state == initial_status.state:
+        # no change in state, i.e. back pressure
+        set_message(
+            job, "Waiting on available resources", code=StatusCode.WAITING_ON_WORKERS
+        )
+
+    elif new_status.state == expected_state:
+        # successful state change to the expected next state
+        if new_status.state == ExecutorState.PREPARING:
+            job.state = State.RUNNING
+        elif job.state != State.RUNNING:
+            # got an ExecutorState that should mean the job.state is RUNNING, but it is not
+            log.warning("state error: got {new_status.state} for job we thought was {job.state}")
+        set_message(job, new_status.state.value.title())
+
+    elif new_status.state == ExecutorState.ERROR:
+        # all transitions can go straight to error
+        mark_job_as_failed(job, new_status.message)
+        api.cleanup(definition)
+
+    else:
+        raise InvalidTransition(
+            f"unexpected state transition of job {job.id} from {initial_status.state} to {new_status.state}: {new_status.message}"
+        )
+
+
+def save_results(job, results):
+    """Extract the results of the execution and update the job accordingly."""
+    # set the final state of the job
+    if results.exit_code != 0:
+        job.state = State.FAILED
+        job.status_message = "Job exited with an error code"
+        job.status_code = StatusCode.NONZERO_EXIT
+    elif results.unmatched_patterns:
+        job.state = State.FAILED
+        job.status_message = "No outputs found matching patterns:\n - {}".format(
+            "\n - ".join(results.unmatched_patterns)
+        )
+        # If the job fails because an output was missing its very useful to
+        # show the user what files were created as often the issue is just a
+        # typo
+
+        # Can we figure these out from job.outputs and project.yaml? Do we do
+        # it here or just in local run?
```

Comment:
Yep, I think we should be able to remove this from the API and handle it as a special case in local run. Haven't quite thought through the details yet though.

Reply:
Ok, should I leave the comment in, or assume we're definitely going to do that and remove it?

Reply:
Might be worth leaving the comment in but changing to "we are planning to remove this ...". Or just leave as is if you want, my comment was just for your info.